grpc-file-transfer/golang/grpc_file_streaming/grpc_stream_io.go

```diff
@@ -0,0 +1,120 @@
+// from the previousChunkHash and the byte array for this chunk.
+func sendMessagesToStream[DataChunkProtoMessage interface{}](
+	payloadNameForLogging string,
+	payload []byte,
+	sendViaStreamFunc func(msg interface{}) error,
+	grpcMsgConstructor func(previousChunkHash string, contentChunk []byte) (*DataChunkProtoMessage, error),
+) error {
+	if len(payload) > kurtosisMaximumFileSizeLimit {
+		return stacktrace.NewError("The files you are trying to send, which are now compressed, exceed or reach " +
+			"100mb. Please reduce the total file size.")
+	}
+
+	var previousChunkHash string
+	hasher := sha1.New()
+	chunkNumber := 0
+	totalChunksNumber := int(math.Ceil(float64(len(payload)) / chunkSize))
+
+	for blockIdx := 0; blockIdx < totalChunksNumber; blockIdx++ {
+		logrus.Debugf("Sending content for %s. Block number %d/%d", payloadNameForLogging, chunkNumber, totalChunksNumber-1)
+
+		var contentChunk []byte
+		payloadStartIdx := blockIdx * chunkSize
+		if blockIdx == totalChunksNumber-1 {
+			// last chunk - copy everything that's left in the payload array
+			contentChunk = make([]byte, len(payload)-payloadStartIdx)
+			copy(contentChunk, payload[payloadStartIdx:])
+		} else {
+			// copy chunkSize bytes from the content array to the contentChunk
+			contentChunk = make([]byte, chunkSize)
+			payloadNextStartIdx := (blockIdx + 1) * chunkSize
+			copy(contentChunk, payload[payloadStartIdx:payloadNextStartIdx])
+		}
+
+		message, err := grpcMsgConstructor(previousChunkHash, contentChunk)
+		if err != nil {
+			return stacktrace.Propagate(err, "An unexpected error occurred assembling GRPC message from data chunk"+
+				" for %s", payloadNameForLogging)
+		}
+
+		err = sendViaStreamFunc(message) // TODO: we can add a retryer here
+		if err != nil {
+			return stacktrace.Propagate(err, "An unexpected error occurred sending '%s'", payloadNameForLogging)
+		}
+
+		hasher.Reset()
+		hasher.Write(contentChunk)
+		previousChunkHash = hex.EncodeToString(hasher.Sum(nil))
+	}
+	return nil
+}
+
+// readMessagesFromStream is a helper function that reads the content of the stream until it returns an io.EOF.
+// It extracts the valuable information (i.e. the byte array and the previous chunk hash) form the generic proto message
+// using the provided grpcMsgExtractor function.
+// It returns a simple byte array corresponding to the assembled payload (concatenation of all chunks)
+func readMessagesFromStream[DataChunkMessageType interface{}](
+	payloadNameForLogging string,
+	readMsgFromStream func(msg interface{}) error,
+	grpcMsgExtractor func(dataChunk *DataChunkMessageType) ([]byte, string, error),
+) ([]byte, error) {
+	var assembledContent []byte
+	var blockIdx int
+	var computedPreviousBlockHash string
+	hasher := sha1.New()
+
+	dataChunk := new(DataChunkMessageType)
+	errorReceivingChunk := readMsgFromStream(dataChunk)
+	for errorReceivingChunk == nil {
```

Comment:
Haha, interesting. I tried it, and dude I find it super confusing. I usually like this pattern, but here it's a bit different because we have to manually allocate the chunk with `dataChunk := new(DataChunkMessageType)` next to the call to `readMsgFromStream`. So if you move this latter call inside the `for ...;...;...` you end up with `dataChunk := new(...)` floating around. even though I agree the current version sliiiightly more verbose, I'll keep it this way to keep our brains sane